### ISSUE TITLE
Feature request: restrict packages to a given date

### ISSUE THREAD

**Issue by Mbodin** [ocaml/opam#5599]
I would like to be able to provide a date to Opam, and make it ignore any package version released after this date. The idea would be that if we run two invocations of Opam (e.g. `opam update` or `opam install <package>`) with the same date as argument, but at different time, then the exact behaviour should be expected (that is, the same version for all packages), even if newest versions of packages are available.

I haven't seen such an option in the documentation. The date to be used for a package could for instance be the published date (e.g. 3rd of March 2021 for Zarith 1.12), as displayed in e.g. https://opam.ocaml.org/packages/zarith/

I would need it for replicability: I would like to get the environment of an artifact, as it were when it was published.

Indeed, I'm currently having issues installing a package, leading to a compilation error on Zarith, which is a dependency of a dependency. I'm going to fill-in an issue about this particular package in the appropriate place once I will understand better what is going on. But from what I've understood now, Zarith's Opam file mentions several dependencies, but without specifying their versions. In such case, Opam fetches the latest possible version of all of them, even if this version has been published after Zarith itself. Unfortunately, one of the newest version of one of the dependency comes with a breaking change, and it makes the compilation fail. It is even more frustrating considering that Zarith isn't the package I tried to install: it is a dependency's dependency.

I feel that being able to fix the date could solve several such reproductibility issues.

**Comment by avsm** [comment]
> I would need it for replicability: I would like to get the environment of an artifact, as it were when it was published.

Have you considered pinning the git revision of the opam repository when you perform the installation?   See, for example, how the [ocaml.org build system](https://github.com/ocaml/ocaml.org/blob/main/Dockerfile#L7) does this.

**Comment by Mbodin** [comment]
I haven't yet ☺ It looks an interesting solution to my particular issue.

I'm not familiar to how Opam works in detail: in the example you gave, it seems that `opam update` is executed from within a copy of the `https://github.com/ocaml/opam-repository` repository at a particular revision. In such case, does any invocation of `opam` base itself on this local instance instead of the remote one?

**Comment by avsm** [comment]
That's right. If you run `opam repo`, it will list  you the remote repositories that opam uses to calculate the packages used.

By default, it tracks `https://opam.ocaml.org`, but you can also pass it a specific Git revision and it will always use that.  See `opam init --help` (for the initial setup), and `opam repo --help` for how to configure remote repositories after you have initialised.

And of course, feel free to ask clarifications here that you can't find answered elsewhere. This sounds like something we should add to the opam FAQ to make it easier to discover.

**Comment by Mbodin** [comment]
Thanks a lot! This helps ☺
